src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/reporting/ndmis/performance/NdmisAppointmentPerformanceReportProcessor.kt

```diff
@@ -0,0 +1,42 @@
+package uk.gov.justice.digital.hmpps.hmppsinterventionsservice.reporting.ndmis.performance
+
+import mu.KLogging
+import net.logstash.logback.argument.StructuredArguments.kv
+import org.springframework.batch.item.ItemProcessor
+import org.springframework.stereotype.Component
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.Referral
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.service.DeliverySessionService
+
+@Component
+class NdmisAppointmentPerformanceReportProcessor(
+  private val deliverySessionService: DeliverySessionService,
+) : ItemProcessor<Referral, List<AppointmentData>> {
+  companion object : KLogging()
+
+  override fun process(referral: Referral): List<AppointmentData>? {
+    logger.debug("processing referral {}", kv("referralId", referral.id))
+
+    if (referral.sentAt == null) throw RuntimeException("invalid referral passed to report processor; referral has not been sent")
+
+    val deliveryAppointments = deliverySessionService.getSessions(referral.id).flatMap { it.appointments }
+    val saaAppointments = referral.supplierAssessment?.appointments ?: emptySet()
+    if (deliveryAppointments.isEmpty() && referral.supplierAssessment?.appointments?.size == 0) {
+      return null
```

Comment:
why is this returning a nullable value? why not just empy list? then we can remove this check entirely

Reply:
Returning null is the same as saying ignore this row and don't include it in the results I believe.  Returning an empty list adds an empty row to the result set.

As an example
[[AppointmentData(referralReference=HD2123AC, referralId=81d754aa-d868-4347-9c0f-50690773014e ...... ..reasonForAppointment=saa)], [], [])

Reply:
got it.

in that case - i would do it like this:

```
    val deliveryAppointments = deliverySessionService.getSessions(referral.id).flatMap { it.appointments }
    val saaAppointments = referral.supplierAssessment?.appointments ?: emptySet()
    
    return (deliveryAppointments + saaAppointments).map { ... }.ifEmpty(null)
```